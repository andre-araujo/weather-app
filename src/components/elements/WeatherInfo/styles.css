.container {
  background: rgba(255, 255, 255, 0.8);
  padding: 10px 8px;
}

.largeText {
  font-size: 16px;
}

.smallText {
  font-size: 16px;
}
