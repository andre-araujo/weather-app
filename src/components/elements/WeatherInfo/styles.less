@import '../../../styles/settings/colors.less';

.container {
  padding: 0 8px;
  display: flex;
  height: 100px;
  overflow: hidden;
  transition: .2s height;
  border: 0;
  width: 100%;
  position: relative;
  box-shadow: 0 2px 6px rgba(0, 0, 0, .2)
}

.openInfo {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
  cursor: pointer;
  background: 0;
  color: transparent;
  user-select: none;
  border: 0;
  z-index: 1;
}

.openInfo:disabled {
  pointer-events: none;
}

.content {
  color: @white;
  padding: 16px 0;
  font-weight: 400;
  text-align: left;
}

.temperature {
  margin-bottom: 30px;
}

.temperature time {
  display: block;
  margin-bottom: 15px;
  text-transform: uppercase;
  font-size: 20px;
}

.toggleTemperature,
.largeText {
  font-size: 23px;
  color: inherit;
  margin-bottom: 16px;
}

.toggleTemperature {
  padding: 0;
  background: none;
  border: 0;
  pointer-events: none;
  outline: none;
}

.smallText {
  font-size: 16px;
  margin-bottom: 4px;
  color: inherit;
}

.icon {
  font-size: 200px;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 16px 50px 0 0;
  width: 58%;
  flex:  0 0 58%;
  color: @white;
  text-align: center;
}

.hidden {
  opacity: 0;
  transition: .2s opacity;
}

.active {
  height: 280px;
  cursor: default;

  .hidden {
    opacity: 1;
  }

  .toggleTemperature {
    pointer-events: all;
  }

  .toggleTemperature:hover {
    cursor: pointer;
    text-shadow: 1px 1px @black10;
  }
}
