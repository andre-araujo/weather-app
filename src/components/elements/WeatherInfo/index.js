import React from 'react';
import { string, number, bool, func, oneOfType } from 'prop-types';

import { CELCIUS, FAHRENHEIT } from '../../../lib/constants';
import WeatherIcon from '../WeatherIcon';
import styles from './styles.less';
import getTemperatureColor from '../../../lib/getTemperatureColor';
import fahrenheitToCelcius from '../../../lib/fahrenheitToCelcius';

function WeatherInfo({
  day,
  humidity,
  pressure,
  temperature,
  temperatureUnit,
  weather,
  weatherCode,
  wind,
  isOpen,
  onClick,
  toggleTemperatureUnit,
}) {
  const convertedTemperature =
    temperatureUnit === CELCIUS ? fahrenheitToCelcius(temperature) : temperature;

  return (
    <div
      className={`${styles.container} ${isOpen ? styles.active : ''}`}
      style={{ backgroundColor: getTemperatureColor(temperature) }}
    >
      {
        !isOpen && (
          <button
            className={styles.openInfo}
            disabled={isOpen}
            onClick={onClick}
            aria-label="Open Info"
            aria-required="true"
          >
            open info
          </button>
        )
      }
      <div className={`${styles.icon} ${styles.hidden}`}>
        <WeatherIcon weatherCode={weatherCode} />
      </div>
      <div className={styles.content}>
        <div className={styles.temperature}>
          <time dateTime={new Date()}>{day}</time>
          <button
            data-test="WeatherInfo_Temperature"
            disabled={!isOpen}
            className={styles.toggleTemperature}
            onClick={toggleTemperatureUnit}
          >
            {convertedTemperature}ยบ{temperatureUnit}
          </button>
        </div>
        <div className={styles.hidden}>
          {weather && <p className={styles.largeText}>{weather}</p>}
          {wind && <p className={styles.smallText}>Wind: {wind}</p>}
          {humidity && <p className={styles.smallText}>humidity: {humidity}</p>}
          {pressure && <p className={styles.smallText}>Pressure: {pressure}</p>}
        </div>
      </div>
    </div>
  );
}

WeatherInfo.defaultProps = {
  day: null,
  temperatureUnit: FAHRENHEIT,
  humidity: null,
  pressure: null,
  temperature: null,
  weather: null,
  weatherCode: null,
  wind: null,
  isOpen: false,
};

WeatherInfo.propTypes = {
  day: string,
  temperatureUnit: string,
  toggleTemperatureUnit: func,
  humidity: string,
  pressure: string,
  temperature: oneOfType([string, number]),
  weather: string,
  weatherCode: string,
  wind: string,
  isOpen: bool,
  onClick: func,
};

export default WeatherInfo;
