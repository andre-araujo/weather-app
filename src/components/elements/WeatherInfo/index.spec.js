import React from 'react';
import { shallow } from 'enzyme';

import WeatherInfo from './index';

let props;

beforeEach(() => {
  props = {
    day: 'Today',
    humidity: '00',
    pressure: '00',
    temperature: '80',
    temperatureUnit: 'F',
    weather: '00',
    weatherCode: '00',
    wind: '00',
    isOpen: true,
    onClick: jest.fn(),
    toggleTemperatureUnit: jest.fn(),
  };
});

describe('WeatherInfo component', () => {
  it('should match snapshot with all props', () => {
    expect(shallow(<WeatherInfo {...props} />)).toMatchSnapshot();
  });

  it('should match snapshot with only required props', () => {
    expect(shallow(<WeatherInfo weatherCode="00" />)).toMatchSnapshot();
  });

  it('should call toggleTemperatureUnit prop when temperature unit is clicked', () => {
    const wrapper = shallow(<WeatherInfo {...props} />);
    wrapper
      .find('[data-test="WeatherInfo_Temperature"]')
      .simulate('click');

    expect(props.toggleTemperatureUnit).toHaveBeenCalledTimes(1);
  });

  it('should converts temperature displayed in FAHRENHEIT', () => {
    props.temperature = '80';
    props.temperatureUnit = 'F';

    const wrapper = shallow(<WeatherInfo {...props} />);
    const displayedTemperature = wrapper
      .find('[data-test="WeatherInfo_Temperature"]')
      .text();

    expect(displayedTemperature).toEqual('80ºF');
  });

  it('should converts temperature displayed in CELCIUS', () => {
    props.temperature = '80';
    props.temperatureUnit = 'C';

    const wrapper = shallow(<WeatherInfo {...props} />);
    const displayedTemperature = wrapper
      .find('[data-test="WeatherInfo_Temperature"]')
      .text();

    expect(displayedTemperature).toEqual('27ºC');
  });
});
