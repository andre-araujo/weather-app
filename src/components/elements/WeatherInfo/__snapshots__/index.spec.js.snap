// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherInfo component should match snapshot with all props 1`] = `
<div
  className="undefined undefined"
  style={
    Object {
      "backgroundColor": "rgba(255, 202, 102, 0.98)",
    }
  }
>
  <div
    className="undefined undefined"
  >
    <WeatherIcon
      weatherCode="00"
    />
  </div>
  <div>
    <div>
      <time
        dateTime={mockConstructor {}}
      >
        Today
      </time>
      <button
        data-test="WeatherInfo_Temperature"
        disabled={false}
        onClick={[Function]}
      >
        80
        ยบ
        F
      </button>
    </div>
    <div>
      <p>
        00
      </p>
      <p>
        Wind: 
        00
      </p>
      <p>
        humidity: 
        00
      </p>
      <p>
        Pressure: 
        00
      </p>
    </div>
  </div>
</div>
`;

exports[`WeatherInfo component should match snapshot with only required props 1`] = `
<div
  className="undefined "
  style={
    Object {
      "backgroundColor": "rgba(100, 100, 255, 0.98)",
    }
  }
>
  <button
    aria-label="Open Info"
    disabled={false}
    onClick={null}
  >
    open info
  </button>
  <div
    className="undefined undefined"
  >
    <WeatherIcon
      weatherCode="00"
    />
  </div>
  <div>
    <div>
      <time
        dateTime={mockConstructor {}}
      />
      <button
        data-test="WeatherInfo_Temperature"
        disabled={true}
        onClick={null}
      >
        ยบ
        F
      </button>
    </div>
    <div />
  </div>
</div>
`;
